### DISCUSSION TITLE
How do I achieve "onLoadMore" for virtualized listbox to enable infinite scrolling? `react-aria-components`

### DISCUSSION THREAD

**Discussion by erangeles** [adobe/react-spectrum#6816] [Q&A] [answered]
I want to achieve infinite loading on virtualized (RAC virtualizer) listbox so when a user gets to the end of the listbox items, I can get a method callback such as `onLoadMore` and trigger loading more items (infinitely). However there is no hook or method to achieve this. I don't want to add a button for this since I want to leverage the scrolling to trigger this. Is there any way to achieve this? 

Example code I'm following https://github.com/adobe/react-spectrum/blob/main/packages/react-aria-components/stories/ListBox.stories.tsx#L242
```
export function VirtualizedListBox({variableHeight}) {
  let sections: {id: string, name: string, children: {id: string, name: string}[]}[] = [];
  for (let s = 0; s < 10; s++) {
    let items: {id: string, name: string}[] = [];
    for (let i = 0; i < 100; i++) {
      const l = (s * 5) + i + 10;
      items.push({id: `item_${s}_${i}`, name: `Section ${s}, Item ${i}${variableHeight ? ' ' + generateRandomString(l, l) : ''}`});
    }
    sections.push({id: `section_${s}`, name: `Section ${s}`, children: items});
  }

  let layout = useMemo(() => {
    return new ListLayout({
      [variableHeight ? 'estimatedRowHeight' : 'rowHeight']: 25,
      estimatedHeadingHeight: 26
    });
  }, [variableHeight]);

  return (
    <Virtualizer layout={layout}>
      <ListBox className={styles.menu} style={{height: 400}} aria-label="virtualized listbox" items={sections}>
        {section => (
          <Section className={styles.group}>
            <Header style={{fontSize: '1.2em'}}>{section.name}</Header>
            <Collection items={section.children}>
              {item => <MyListBoxItem>{item.name}</MyListBoxItem>}
            </Collection>
          </Section>
        )}
      </ListBox>
    </Virtualizer>
  );
}

VirtualizedListBox.args = {
  variableHeight: false
};
```

**Comment by snowystinger** [comment, marked as answer]
We'll be adding documentation for RAC virtualization as the API settles. In the meantime, it's probably easiest to look at our code as an example https://github.com/adobe/react-spectrum/blob/617ad2ed3c3ee46715ca505a315cbc2b945b1f4e/packages/react-aria-components/stories/Table.stories.tsx#L792

We appreciate your patience as we continue to improve this set of features.

**Reply by stevegreco** [reply]
@snowystinger Any update on when this might be documented/stable?  We would love to use this hook in our codebase, as this solves some problems with using this functionality in conjunction with the ListBox.

**Reply by snowystinger** [reply]
No update on the documentation right now, but we've been using the APIs for a while now, so the chance that they'd change is substantially lower than when I originally wrote this comment. Will bring up with the team for discussion, but I'd refer to our code as an example for now.

**Reply by stevegreco** [reply]
Thanks!  The example is very helpful.  We find this hook really useful and are looking forward to using it.

**Reply by selrond** [reply]
@snowystinger any update now? Virtualizer has been in GA [for some time now](https://react-spectrum.adobe.com/releases/2025-03-05.html), and the load more functionality is quite related

**Reply by snowystinger** [reply]
Hey! Thanks for the interest. We're actually working on load more work right now https://github.com/adobe/react-spectrum/pull/8431#issue-3164395066

There's a preview of the docs in that PR.

We are also in the midst of updating all of our docs, so we'll hopefully have easier to understand and better grouped information on these topics soon.

**Reply by stevegreco** [reply]
> Hey! Thanks for the interest. We're actually working on load more work right now https://github.com/adobe/react-spectrum/pull/8431#issue-3164395066
> 
> There's a preview of the docs in that PR.
> 
> We are also in the midst of updating all of our docs, so we'll hopefully have easier to understand and better grouped information on these topics soon.

This is great! This is going to solve so many problems for our adopters! Thank you!

**Comment by itay-perry** [comment]
Is there an example in the docs that combines the `ListBox`, `ListBoxLoadMoreItem` and the `Virtualizer` together? 🙏🏽 And then we could use both of these very powerful features together 🚀

Update ✨ Found this in the test files: https://github.com/adobe/react-spectrum/blob/9da8c73d96bc7bef32de5440ad1877ba03da7dfe/packages/react-aria-components/test/ListBox.test.js#L1668-L1692

**Reply by itay-perry** [reply]
I really think this should be added to the docs, it's extremely valuable :)
